File: examples/rkky.cpp
// RKKY example from https://mumax.github.io/examples.html
#include "field_terms/micro/demag_field.hpp"
#include "field_terms/micro/exchange_field.hpp"
#include "field_terms/micro/rkky_exchange_field.hpp"
#include "integrators/llg_integrator.hpp"
#include "util/arg_parser.hpp"
#include <cmath>

using namespace magnumafcpp;

int main(int argc, char** argv) {
    const auto [outdir, posargs] = ArgParser(argc, argv).outdir_posargs;

    // Parameter initialization
    const int nx = 10, ny = 10, nz = 2;
    const double dx = 1e-9;

    const double Ms = 1e6;
    const double A = 10e-12;
    const double RKKY = -1e-3 * dx;

    // Generating Objects
    Mesh mesh{nx, ny, nz, dx, dx, dx};

    // Initial magnetic field
    af::array m = af::constant(0.0, mesh::dims_v(mesh), f64);
    m(af::span, af::span, af::span, 0) = 1.;
    State state(mesh, Ms, m);
    state.write_vti(outdir / "minit");
    af::array rkkyvals = af::constant(RKKY / 2., mesh::dims_v(mesh), f64);
    af::array exchvals = af::constant(A, mesh::dims_v(mesh), f64);
    auto rkky = uptr_FieldTerm(new RKKYExchangeField(RKKY_values(rkkyvals), Exchange_values(exchvals), mesh));

    auto demag = uptr_FieldTerm(new DemagField(mesh, true, true, 0));
    LLGIntegrator Llg(1, {std::move(demag), std::move(rkky)});

    std::ofstream stream(outdir / "m.dat");
    stream.precision(12);

    std::ofstream streamE(outdir / "E.dat");
    streamE.precision(12);

    std::vector<double> vecE;

    for (int i = 0; i < 360; i++) {
        const double mix = std::cos(i * M_PI / 180.);
        const double miy = std::sin(i * M_PI / 180.);
        state.m(af::span, af::span, 1, 0) = mix;
        state.m(af::span, af::span, 1, 1) = miy;

        double E = Llg.E(state);
        vecE.push_back(E);
        std::cout << "i = " << i << ", E= " << E << std::endl;
        streamE << i << "\t" << E << std::endl;
        stream << state << std::endl;
        state.write_vti(outdir / ("m_field_" + std::to_string(i)));
    }
    stream.close();
    streamE.close();

    double maxval = *std::max_element(std::begin(vecE), std::end(vecE));
    double minval = *std::min_element(std::begin(vecE), std::end(vecE));

    std::cout << "maxval = " << maxval << std::endl;
    std::cout << "minval = " << minval << std::endl;
    std::cout << "Diff maxval minval  = " << maxval - minval << std::endl;

    return 0;
}

File: python/examples/rkky.py
#!/usr/bin/python3
# RKKY example from https://mumax.github.io/examples.html

import numpy as np
import arrayfire as af
from magnumaf import *
import sys
import time

args = parse()
af.info()

# Discretization
nx, ny, nz = 10, 10, 2
dx, dy, dz = 1e-9, 1e-9, 1e-9

# Material parameters
Ms = 1e6
A = 10e-12
RKKY = -1e-3 * dz # TODO move to source

# Initial magnetization configuration
m0 = af.constant(0.0, nx, ny, nz, 3, dtype=af.Dtype.f64)
m0[:, :, :, 0] = 1.

# Creating objects
mesh = Mesh(nx, ny, nz, dx, dy, dz)
state = State(mesh, Ms, m = m0)

demag = DemagField(mesh, verbose = True, caching = False, nthreads = 6)

rkky_values = af.constant(RKKY/2., nx, ny, nz, 1, dtype=af.Dtype.f64)
exch_values = af.constant(A, nx, ny, nz, 1, dtype=af.Dtype.f64)
rkky_indices = af.constant(0, nx, ny, nz, 1, dtype=af.Dtype.u32)
rkkyexch = RKKYExchangeField(rkky_values, exch_values, mesh, rkky_indices)

llg = LLGIntegrator(alpha = 1, terms = [demag, rkkyexch])

E = []
print("Start rotating")
stream = open(args.outdir + "m.dat", "w")
timer = time.time()
for i in range(0, 360):
    mix = np.cos(i * np.pi/180.);
    miy = np.sin(i * np.pi/180.);
    m = state.m
    m[:, :, 1, 0] = mix
    m[:, :, 1, 1] = miy
    state.m = m
    E.append( llg.Eeff_in_J(state) )
    print("angle=[°]", i, " E=", E[-1])
    mean = af.mean(af.mean(af.mean(state.m, dim=0), dim=1), dim=2)
    stream.write("%d, %e, %e, %e, %e\n" %(i, E[-1], mean[0, 0, 0, 0].scalar(), mean[0, 0, 0, 1].scalar(), mean[0, 0, 0, 2].scalar()))
print("fullrotation in ", time.time() - timer, "[s]")
stream.close()

print("E diff =", max(E) - min(E), "[J], should be around 2.1e-19 (from mumax3 plot)")

# plotting data with gnuplot
from os import system
system('gnuplot -e "\
    set terminal pdf;\
    set output \'' + args.outdir + 'm.pdf\';\
    set xlabel \'angle [°]\';\
    set ylabel \'E\';\
    p \'' + args.outdir + '/m.dat\' u 1:2 w l t \'E\';\
    set ylabel \'<m>\';\
    p \'' + args.outdir + '/m.dat\' u 1:3 w l t \'<m_x>\',\
    \'\' u 1:4 w l t \'<m_y>\',\
    \'\' u 1:5 w l t \'<m_z>\';\
"')

# show pdf with evince
system('evince ' + args.outdir +'m.pdf')
